### PR TITLE
Bug 1950978: ingressclass: Handle ingresscontroller deletion

### PR DESCRIPTION
**Pull request by Miciah**
#### ingressclass: Handle ingresscontroller deletion

If the ingresscontroller corresponding to an ingressclass does not exist (for example, because the ingresscontroller was deleted), delete the ingressclass.

* `manifests/00-cluster-role.yaml`: Allow the operator to delete ingressclasses.
* `pkg/manifests/bindata.go`: Regenerate.
* `pkg/operator/controller/ingressclass/controller.go` (`Reconcile`): Pass the ingresscontroller's namespaced name to `ensureIngressClass`.
* `pkg/operator/controller/ingressclass/ingressclass.go` (`ensureIngressClass`): Change ingresscontroller name parameter type from string to namespaced-name.  Check if the ingresscontroller exists and is not marked for deletion, and pass a corresponding Boolean value to `desiredIngressClass`.
(`desiredIngressClass`): Return `false, nil` if the ingresscontroller does not exist.
* `test/e2e/operator_test.go` (`TestCustomIngressClass`): New test.  Verify that the ingressclass controller creates an ingressclass for a custom ingresscontroller and deletes the ingressclass when the ingresscontroller is deleted.

#### ingressclass: Capitalize names consistently

* `pkg/operator/controller/ingressclass/ingressclass.go` (`ensureIngressClass`, `desiredIngressClass`, `currentIngressClass`, `updateIngressClass`): Capitalize "ingressclasses" as "ingressClasses" and "ingresscontrollerName" as "ingressControllerName".
(`ingressclassChanged`): Rename...
(`ingressClassChanged): ...to this.`

#### ingressclass: Add unit tests

* `pkg/operator/controller/ingressclass/ingressclass_test.go`: New file.
(`TestDesiredIngressClass`): New test.  Verify that `desiredIngressClass` behaves as expected.
(`TestIngressClassChanged`): New test.  Verify that `ingressClassChanged` behaves as expected.

#### `TestDefaultIngressClass`: Omit namespace in logs

* `test/e2e/operator_test.go` (`TestDefaultIngressClass`): Omit the ingressclass's namespace in log messages.  The namespace is always empty because ingressclass is cluster-scoped.

### CODE REVIEW

**Comment by sgreene570** [comment]
e2e-aws-operator failures _might_ have been flakes.
/test e2e-aws-operator

File: pkg/operator/controller/ingressclass/ingressclass.go [hunk 1]
```diff
@@ -59,7 +73,11 @@ func (r *reconciler) ensureIngressClass(ingresscontrollerName string, ingresscla
 
 // desiredIngressClass returns a Boolean indicating whether an IngressClass
 // is desired, as well as the IngressClass if one is desired.
-func desiredIngressClass(ingresscontrollerName string, ingressclasses []networkingv1.IngressClass) (bool, *networkingv1.IngressClass) {
+func desiredIngressClass(haveIngressController bool, ingresscontrollerName string, ingressclasses []networkingv1.IngressClass) (bool, *networkingv1.IngressClass) {
+	if !haveIngressController {
```
Comment (sgreene570) [inline pkg/operator/controller/ingressclass/ingressclass.go:77]: Doesn't have to happen in this PR, but should we consider adding some unit test coverage for `desiredIngressClass` and `ingressclassChanged`? Speaking of which, at some point we should change `ingressclassChanged` to `ingressClassChanged` as well :grin:

**Comment by sgreene570** [comment]
/hold
This PR is consistently failing `TestDefaultIngressClass` & `TestCustomIngressClass`.

File: pkg/operator/controller/ingressclass/ingressclass.go [hunk 2]
```diff
@@ -15,13 +15,27 @@ import (
...
-// with the given name.  Returns a Boolean indicating whether the IngressClass
-// exists, the current IngressClass if it does exist, and an error value.
+// with the given name if the IngressController exists, or ensures that such an
+// IngressClass doesn't exist if the IngressController does not exist.  Returns
+// a Boolean indicating whether the IngressClass exists, the current
+// IngressClass if it does exist, and an error value.
 func (r *reconciler) ensureIngressClass(ingresscontrollerName string, ingressclasses []networkingv1.IngressClass) (bool, *networkingv1.IngressClass, error) {
-	want, desired := desiredIngressClass(ingresscontrollerName, ingressclasses)
+	haveIngressController := false
+	icName := types.NamespacedName{Name: ingresscontrollerName}
```
Comment (Miciah) [inline pkg/operator/controller/ingressclass/ingressclass.go:28]: Ugh, apparently I forgot ingresscontrollers were namespaced.

File: pkg/operator/controller/ingressclass/ingressclass.go (hunk 1 above)
Comment (Miciah) [reply pkg/operator/controller/ingressclass/ingressclass.go:76]: Added unit tests and fixed capitalization.  Thanks!

File: test/e2e/operator_test.go [hunk 3]
```diff
@@ -192,24 +192,86 @@ func TestDefaultIngressClass(t *testing.T) {
...
 	if err != nil {
-		t.Fatalf("failed to observe recreated ingressclass %q: %v", name, err)
+		t.Fatalf("failed to observe recreated ingressclass %q: %v", name.Name, err)
+	}
+}
+
+// TestCustomIngressClass verifies that the ingressclass controller creates an
+// ingressclass for a custom ingresscontroller and deletes the ingressclass if
+// the ingresscontroller is deleted.
+func TestCustomIngressClass(t *testing.T) {
```
Comment (sgreene570) [inline test/e2e/operator_test.go:219]: Does the operator have deletion privileges for ingress classes?

File: pkg/operator/controller/ingressclass/ingressclass_test.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+			// annotation.
+			expectIngressClass: makeIngressClass("custom", false),
+		},
+		{
+			description:           "default ingresscontroller when no default ingressclass exists",
+			haveIngressController: true,
+			ingressControllerName: "default",
+			ingressClasses:        []networkingv1.IngressClass{},
+			expectWant:            true,
+			// TODO This test case expects the default ingressclass
```
Comment (sgreene570) [inline pkg/operator/controller/ingressclass/ingressclass_test.go:92]: Thanks for nothing this!

**Review by sgreene570 — COMMENTED** [review]
unit test coverage looks great!

I think the only thing this PR is missing is some RBAC changes :thinking:

File: test/e2e/operator_test.go (hunk 3 above)
Comment (Miciah) [reply test/e2e/operator_test.go:219]: Sigh.  Fixed.

**Comment by sgreene570** [comment]
e2e-aws-operator has passed
/hold cancel
/retest

**Pull request merged**